scrapy/http/request/__init__.py

```diff
@@ -103,3 +104,32 @@ def replace(self, *args, **kwargs):
             kwargs.setdefault(x, getattr(self, x))
         cls = kwargs.pop('cls', self.__class__)
         return cls(*args, **kwargs)
+
+    @classmethod
+    def from_curl(cls, curl_command, ignore_unknown_options=True, **kwargs):
+        """Create a Request object from a string containing a `cURL
+        <https://curl.haxx.se/>`_ command. It populates the HTTP method, the
+        URL, the headers, the cookies and the body. It accepts the same
+        arguments as the :class:`Request` class, taking preference and
+        overriding the values of the same arguments contained in the cURL
+        command.
+
+        Unrecognized options are ignored by default. To raise an error when
+        finding unknown options call this method by passing
+        ``ignore_unknown_options=False``.
+
+        .. caution:: Enabled
+                     :ref:`downloader middlewares <topics-downloader-middleware>`
+                     and
+                     :ref:`spider middlewares <topics-spider-middleware>`,
+                     such as
+                     :class:`~scrapy.downloadermiddlewares.defaultheaders.DefaultHeadersMiddleware`,
+                     :class:`~scrapy.downloadermiddlewares.useragent.UserAgentMiddleware`,
+                     or
+                     :class:`~scrapy.downloadermiddlewares.httpcompression.HttpCompressionMiddleware`,
+                     may modify the :class:`~scrapy.http.Request` object.
+
+       """
+        request_kwargs = curl_to_request_kwargs(curl_command, ignore_unknown_options)
+        request_kwargs.update(kwargs)
+        return Request(**request_kwargs)
```

Comment:
I wonder if it should use cls instead of Request, not sure about this. Opinions?

Reply:
Oh, that's right... It’s not as trivial as it seems. Suppose someone uses from_curl from JSONRequest:

`JSONRequest.from_curl('curl example.org')`.

Using `Request` the result would be:

```
{'_encoding': 'utf-8',
 'method': 'GET',
 '_url': 'http://example.org',
 '_body': b'',
 'priority': 0,
 'callback': None,
 'errback': None,
 'cookies': {},
 'headers': {},
 'dont_filter': False,
 '_meta': None,
 '_cb_kwargs': None,
 'flags': []}
```

Using `cls` the result would be:

```
{'_dumps_kwargs': {'sort_keys': True},
 '_encoding': 'utf-8',
 'method': 'GET',
 '_url': 'http://example.org',
 '_body': b'',
 'priority': 0,
 'callback': None,
 'errback': None,
 'cookies': {},
 'headers': {b'Content-Type': b'application/json',
  b'Accept': b'application/json, text/javascript, */*; q=0.01'},
 'dont_filter': False,
 '_meta': None,
 '_cb_kwargs': None,
 'flags': []}
```

(The same happens with `XmlRpcRequest`)
I think that the first approach could be better, because it respects the curl command, but I understand that for some users could be counter-intuitive.. seems a UX dilema. In any case, we should probably add something to the docstring. What do you think? @Gallaecio any thoughs?

Reply:
If we are not using `cls`, maybe we should reimplement the new class method in our built-in `Request` subclasses to raise a `NotImplementedError` exception.

Reply:
I think it's the same situation as with middlewares: scrapy should try to create the request with the inputs specified, and then if they are overridden by the class itself, or by the middlewares, you just need to be prepared to that it might happen and not expect 100% conformance.

Reply:
It seems a valid argument. We could use `cls` and extend the "caution" message regarding the middlewares adding that if it's used from `Request` subclasses they could add some modifications to the request.

Reply:
I add a commit with that idea. @kmike @Gallaecio tell me what you think. :)